engine/src/metrics.rs

```diff
@@ -0,0 +1,196 @@
+    /// buffer cache before the write is considered complete. If this
+    /// flag is true, writes will be slower.
+    ///
+    /// # Errors
+    /// Return `EngineError::TableNotFound` if the given table does not exist
+    /// Return `EngineError` if met some errors
+    fn write_batch(&self, wr_ops: Vec<WriteOperation<'_>>, sync: bool) -> Result<(), EngineError> {
+        let start = Instant::now();
+        let batch_size = wr_ops.len().to_string();
+        let res = self.engine.write_batch(wr_ops, sync);
+        metrics().engine_write_batch_duration_seconds.record(
+            start.elapsed().as_secs(),
+            &[
+                KeyValue::new("batch_size", batch_size),
+                KeyValue::new("sync", u8::from(sync).to_string()),
+            ],
+        );
+        res
+    }
+
+    /// Get a snapshot of the current state of the database
+    ///
+    /// # Errors
+    /// Return `EngineError` if met some errors when creating the snapshot
+    fn get_snapshot(
+        &self,
+        path: impl AsRef<Path>,
+        tables: &[&'static str],
+    ) -> Result<Self::Snapshot, EngineError> {
+        self.engine.get_snapshot(path, tables).map(Layer::new)
+    }
+
+    /// Apply a snapshot to the database
+    ///
+    /// # Errors
+    /// Return `EngineError` if met some errors when applying the snapshot
+    async fn apply_snapshot(
+        &self,
+        snapshot: Self::Snapshot,
+        tables: &[&'static str],
+    ) -> Result<(), EngineError> {
+        let start = Instant::now();
+        let res = self.engine.apply_snapshot(snapshot.engine, tables).await;
+        metrics()
+            .engine_apply_snapshot_duration_seconds
+            .record(start.elapsed().as_secs(), &[]);
+        res
+    }
+
+    /// Get the cached size of the engine (Measured in bytes)
+    fn estimated_file_size(&self) -> u64 {
+        self.engine.estimated_file_size()
+    }
+
+    /// Get the file size of the engine (Measured in bytes)
+    fn file_size(&self) -> Result<u64, EngineError> {
+        self.engine.file_size()
+    }
+}
+
+#[async_trait]
+impl<E> SnapshotApi for Layer<E>
+where
+    E: SnapshotApi,
+{
+    /// Get the size of the snapshot
+    fn size(&self) -> u64 {
+        self.engine.size()
+    }
+
+    /// Rewind the snapshot to the beginning
+    fn rewind(&mut self) -> io::Result<()> {
+        self.engine.rewind()
+    }
+
+    /// Pull some bytes of the snapshot to the given uninitialized buffer
+    async fn read_buf(&mut self, buf: &mut BytesMut) -> io::Result<()> {
+        self.engine.read_buf(buf).await
+    }
+
+    /// Write the given buffer to the snapshot
+    async fn write_all(&mut self, buf: Bytes) -> io::Result<()> {
+        self.engine.write_all(buf).await
+    }
+
+    /// Clean files of current snapshot
+    async fn clean(&mut self) -> io::Result<()> {
+        self.engine.clean().await
+    }
+}
+
+/// The backend metrics
+static METRICS: OnceLock<Metrics> = OnceLock::new();
+
```

Comment:
Why don't we need a `METRICS_METER` here?

Reply:
fixed